src/app/games/diaballik/DiaballikMoveGenerator.ts

```diff
@@ -5,61 +5,146 @@ import { DiaballikNode, DiaballikRules } from './DiaballikRules';
 import { Player, PlayerOrNone } from 'src/app/jscaip/Player';
 import { Coord } from 'src/app/jscaip/Coord';
 import { Direction, Orthogonal } from 'src/app/jscaip/Direction';
-import { MGPFallible } from 'src/app/utils/MGPFallible';
 import { MGPOptional } from 'src/app/utils/MGPOptional';
+import { Utils } from 'src/app/utils/utils';
+import { MGPFallible } from 'src/app/utils/MGPFallible';
+
+class DiaballikMoveInConstruction {
+
+    public readonly hasPass: boolean;
+    public readonly translations: number;
+
+    public constructor(public readonly subMoves: DiaballikSubMove[],
+                       public readonly stateAfterSubMoves: DiaballikState) {
+        Utils.assert(this.subMoves.length <= 3, 'DiaballikMoveInConstruction can have at most 3 submoves');
+        let hasPass: boolean = false;
+        let translations: number = 0;
+        for (const subMove of subMoves) {
+            if (subMove instanceof DiaballikPass) {
+                Utils.assert(hasPass === false, 'DiaballikMoveInConstruction can have at moste one pass');
+                hasPass = true;
+            } else {
+                Utils.assert(translations < 2, 'DiaballikMoveInConstruction can have at most two translations');
+                translations++;
+            }
+        }
+        this.hasPass = hasPass;
+        this.translations = translations;
+    }
+
+    public addIfLegal(subMove: DiaballikSubMove, listToAddTo: DiaballikMoveInConstruction[]): void {
+        const legality: MGPFallible<DiaballikState> = DiaballikRules.get().isLegalSubMove(this.stateAfterSubMoves, subMove);
+        if (legality.isSuccess()) {
+            const newMoveInConstruction: DiaballikMoveInConstruction =
+                new DiaballikMoveInConstruction(this.subMoves.concat([subMove]), legality.get());
+            listToAddTo.push(newMoveInConstruction);
+        }
+    }
+
+    public getPassEnd(): MGPOptional<Coord> {
+        for (const subMove of this.subMoves) {
+            if (subMove instanceof DiaballikPass) {
+                return MGPOptional.of(subMove.getEnd());
+            }
+        }
+        /* istanbul ignore next */
+        // Unreachable because getPassEnd is only called when there is a pass
+        return MGPOptional.empty();
+    }
+
+    public getPreviousTranslation(): MGPOptional<DiaballikTranslation> {
+        for (const subMove of this.subMoves) {
+            if (subMove instanceof DiaballikTranslation) {
+                return MGPOptional.of(subMove);
+            }
+        }
+        /* istanbul ignore next */
+        // Unreachable because getPreviousTranslation is only called when there is a previous translation
+        return MGPOptional.empty();
```

Comment:
alors soit tu fait un test "should throw then" pour couvrir soit tu reforumule d'une façon couvrable :p

Reply:
mais ce test n'est pas écrivable sans exporter la classe, donc en polluant nos exports avec des trucs internes !